### PR TITLE
[Ray DAG][Serve Deployment Graph] Remove double json.dumps in DAGNode

### LINKED ISSUE #23314: [Feature] Fix double dumped JSON in serve dagnode
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

We should remove the extra hop of `json.dumps()` at DAGNode base implementation. It makes the yaml super weird.

### Use case

.

### Related issues

.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by jiaodong**
## Why are these changes needed?

Previously we have double dump behavior that makes json serde not human readable or friendly, but it's required given `DAGDriver` takes `dag_node_json` as first arg and it will appear in YAML.

This PR removes extra `json.dumps()` in encoder path, eliminated and simplified most encoder / object_hooks that are not needed in the first place to make everything simpler again.

Sample YAML now for a complex DAG: https://gist.github.com/jiaodong/32991771e9d78c35767eb24ed73f8236

### Next Step

We're pretty close to have a better minimal JSON representation of the whole dag after this. I might include in this PR or separate one.

## Related issue number

Closes #23314

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/pipeline/json_serde.py [hunk 1]
```diff
@@ -142,28 +142,22 @@ def dagnode_from_json(input_json: Any) -> Union[DAGNode, RayServeHandle, Any]:
         )
     # Deserialize DAGNode type
```
Comment (jiaodong) [inline python/ray/serve/pipeline/json_serde.py:143]: In my next PR to introduce minimal JSON, I merged these into a simple name -> cls dispatch table. It's not in this PR for simplicity.

**Review by edoakes — APPROVED** [review]
Going in the right direction!

I think in the future we can get rid of the requirement to JSON-serialize any of the node types entirely if we can simplify the `DAGHandle` to not wrap DAG node types directly, but we can discuss that further offline.

We should definitely make it such that the example config you posted would not redundantly include all of the deployment options for each node.

**Comment by edoakes** [comment]
Tests flaky/failing on master

**Comment by jiaodong** [comment]
> Going in the right direction!
> 
> 
> 
> I think in the future we can get rid of the requirement to JSON-serialize any of the node types entirely if we can simplify the `DAGHandle` to not wrap DAG node types directly, but we can discuss that further offline.
> 
> 
> 
> We should definitely make it such that the example config you posted would not redundantly include all of the deployment options for each node.

That's the in progress PR 😄 i think we are very close to it now.

**Pull request merged**